File: school/www/student_registration.js
// student_registration.js
// Handles loading classes based on selected school

frappe.ready(function() {
    console.log('Student Registration JS loaded');

    const schoolSelect = document.getElementById('school');
    const classSelect = document.getElementById('student_class');

    // Ensure class select starts disabled
    if (classSelect) {
        classSelect.innerHTML = '<option value="">Select school first</option>';
        classSelect.disabled = true;
    }

    if (schoolSelect) {
        schoolSelect.addEventListener('change', function() {
            const selectedSchool = this.value;
            console.log('School changed to:', selectedSchool);

            if (classSelect) {
                classSelect.innerHTML = '<option value="">Select class</option>';
                classSelect.disabled = true;
                classSelect.value = '';
            }

            if (selectedSchool) {
                loadClassesBySchool(selectedSchool);
            } else {
                if (classSelect) {
                    classSelect.innerHTML = '<option value="">Select school first</option>';
                    classSelect.disabled = true;
                }
            }
        });
    }

    // Clear error state on interaction
    document.querySelectorAll('input, select').forEach(function(field) {
        field.addEventListener('change', function(e) { e.target.classList.remove('error'); });
        field.addEventListener('input',  function(e) { e.target.classList.remove('error'); });
    });
});

// ---------------------------------------------------------------------------
// Class loading
// ---------------------------------------------------------------------------

async function loadClassesBySchool(school) {
    const classSelect = document.getElementById('student_class');

    if (!classSelect || !school) {
        if (classSelect) {
            classSelect.innerHTML = '<option value="">Select school first</option>';
            classSelect.disabled = true;
        }
        return;
    }

    classSelect.innerHTML = '<option value="">Loading classes\u2026</option>';
    classSelect.disabled = true;

    try {
        const csrf = getCsrfToken();
        const resp = await fetch(
            '/api/method/school.www.student_registration.get_classes_by_school',
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Frappe-CSRF-Token': csrf
                },
                credentials: 'include',
                body: JSON.stringify({ school: school })
            }
        );

        const r = await resp.json();

        if (r.message && r.message.length > 0) {
            classSelect.innerHTML = '<option value="">Select class</option>';
            r.message.forEach(function(cls) {
                const option = document.createElement('option');
                option.value = cls.name;
                option.textContent = cls.name;
                classSelect.appendChild(option);
            });
            classSelect.disabled = false;
        } else {
            classSelect.innerHTML = '<option value="">No classes available</option>';
            classSelect.disabled = true;
        }
    } catch (e) {
        console.error('Error loading classes:', e);
        classSelect.innerHTML = '<option value="">Error loading classes</option>';
        classSelect.disabled = true;
    }
}

// ---------------------------------------------------------------------------
// CSRF helper
// ---------------------------------------------------------------------------

function getCsrfToken() {
    return (
        document.cookie
            .split(';')
            .map(function(c) { return c.trim(); })
            .find(function(c) { return c.startsWith('csrf_token='); }) || ''
    ).replace('csrf_token=', '');
}

// ---------------------------------------------------------------------------
// Guardian fields toggle
// ---------------------------------------------------------------------------

function toggleGuardianFields() {
    const guardianIs = document.getElementById('if_guardian_is');
    const guardianFields = document.querySelectorAll('.g-other');

    if (guardianIs && guardianIs.value === 'Other') {
        guardianFields.forEach(function(f) { f.style.display = 'block'; });
    } else {
        guardianFields.forEach(function(f) { f.style.display = 'none'; });
    }
}

// ---------------------------------------------------------------------------
// Field helper
// ---------------------------------------------------------------------------

function getFieldValue(fieldId) {
    const field = document.getElementById(fieldId);
    return field ? field.value.trim() : '';
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

function validateForm() {
    // Only validate step-1 required fields when step 1 is active
    const step1Active = document.getElementById('step1') &&
                        document.getElementById('step1').classList.contains('active');

    const requiredFields = ['school', 'first_name', 'last_name', 'student_type'];

    // Only require student_class if the select is enabled (classes have loaded)
    const classSelect = document.getElementById('student_class');
    if (classSelect && !classSelect.disabled) {
        requiredFields.push('student_class');
    }

    let isValid = true;

    requiredFields.forEach(function(fieldId) {
        const field = document.getElementById(fieldId);
        if (field) {
            if (!field.value) {
                field.classList.add('error');
                isValid = false;
            } else {
                field.classList.remove('error');
            }
        }
    });

    // Payment fields only required when billing is enabled (step 4)
    const billEnabled = document.body.getAttribute('data-bill-enabled') === 'true';
    if (billEnabled) {
        ['account', 'payment_method'].forEach(function(fieldId) {
            const field = document.getElementById(fieldId);
            if (field) {
                if (!field.value) {
                    field.classList.add('error');
                    isValid = false;
                } else {
                    field.classList.remove('error');
                }
            }
        });
    }

    return isValid;
}

// ---------------------------------------------------------------------------
// Step navigation
// ---------------------------------------------------------------------------

function goToStep(step) {
    const currentStep = document.querySelector('.form-step.active');
    const nextStep    = document.getElementById('step' + step);

    if (!currentStep || !nextStep) return;

    const currentNum = parseInt(currentStep.id.replace('step', ''), 10);

    // Only validate when moving forward from step 1
    if (step > currentNum && currentNum === 1 && !validateForm()) {
        showError('Please fill in all required fields (School, First Name, Last Name, Student Type, and Class).');
        return;
    }

    currentStep.classList.remove('active');
    nextStep.classList.add('active');

    document.querySelectorAll('.step-tab').forEach(function(tab, index) {
        if (index + 1 === step) {
            tab.classList.add('active');
            tab.classList.remove('done');
        } else if (index + 1 < step) {
            tab.classList.add('done');
            tab.classList.remove('active');
        } else {
            tab.classList.remove('active', 'done');
        }
    });

    if (step === 5) buildReview();
    window.scrollTo({ top: 0, behavior: 'smooth' });
}

// ---------------------------------------------------------------------------
// Review builder  (fixed — no stray Chinese characters)
// ---------------------------------------------------------------------------

function escapeHtml(text) {
    const div = document.createElement('div');
    div.textContent = text;
    return div.innerHTML;
}

function buildReview() {
    const reviewContent = document.getElementById('reviewContent');
    if (!reviewContent) return;

    const fields = {
        'School':        getFieldValue('school'),
        'Student Name':  [getFieldValue('first_name'), getFieldValue('second_name'), getFieldValue('last_name')].filter(Boolean).join(' '),
        'Class':         getFieldValue('student_class'),
        'Student Type':  getFieldValue('student_type'),
        'Gender':        getFieldValue('gender'),
        'Date of Birth': getFieldValue('date_of_birth'),
        'Father Name':   getFieldValue('father_name'),
        'Mother Name':   getFieldValue('mother_name'),
        'Father Phone':  getFieldValue('phone_number'),
        'Mother Phone':  getFieldValue('mother_phone'),
        'Guardian Is':   getFieldValue('if_guardian_is'),
        'Guardian Name': getFieldValue('guardian_name'),
        'Address':       getFieldValue('current_address')
    };

    let html = '<table style="width:100%;border-collapse:collapse">';
    let i = 0;

    for (const [label, value] of Object.entries(fields)) {
        if (!value) continue;
        const bg = i % 2 ? '#fff' : '#fdfbf7';
        html += '<tr style="background:' + bg + '">' +
            '<td style="padding:9px 14px;font-size:11px;font-weight:600;text-transform:uppercase;' +
                'color:#7f8c8d;width:40%;border-bottom:1px solid #f9f5ee">' + escapeHtml(label) + '<\/td>' +
            '<td style="padding:9px 14px;font-size:14px;border-bottom:1px solid #f9f5ee">' + escapeHtml(value) + '<\/td>' +
            '<\/tr>';
        i++;
    }

    html += '<\/table>';
    reviewContent.innerHTML = html;
}

// ---------------------------------------------------------------------------
// Error display
// ---------------------------------------------------------------------------

function showError(message) {
    const errorDiv = document.getElementById('globalError');
    if (errorDiv) {
        errorDiv.textContent = message;
        errorDiv.classList.add('show');
        errorDiv.scrollIntoView({ behavior: 'smooth', block: 'center' });
        setTimeout(function() { errorDiv.classList.remove('show'); }, 5000);
    } else {
        alert(message);
    }
}

// ---------------------------------------------------------------------------
// CSRF helper – fetches a fresh token from Frappe for guest sessions
// ---------------------------------------------------------------------------
async function getFrappeCSRF() {
    // 1. Try the cookie first (works for logged-in users)
    const fromCookie = document.cookie
        .split(';')
        .map(function(c) { return c.trim(); })
        .find(function(c) { return c.startsWith('csrf_token='); });

    if (fromCookie) {
        const token = decodeURIComponent(fromCookie.split('=')[1]);
        // Frappe sets 'Fetch' as the guest placeholder – means we need a real one
        if (token && token !== 'Fetch') return token;
    }

    // 2. For guest sessions, fetch a real token from Frappe
    try {
        const r = await fetch('/api/method/frappe.auth.get_logged_user', {
            method: 'GET',
            credentials: 'same-origin'
        });
        // After this request Frappe sets the csrf_token cookie properly
        const fromCookieAfter = document.cookie
            .split(';')
            .map(function(c) { return c.trim(); })
            .find(function(c) { return c.startsWith('csrf_token='); });
        if (fromCookieAfter) {
            return decodeURIComponent(fromCookieAfter.split('=')[1]);
        }
    } catch (e) {
        console.warn('Could not refresh CSRF token:', e);
    }

    return '';
}

// ---------------------------------------------------------------------------
// Submission
// ---------------------------------------------------------------------------
async function submitRegistration() {
    if (!validateForm()) {
        showError('Please fill in all required fields.');
        return false;
    }

    const declaration = document.getElementById('declaration');
    if (!declaration || !declaration.checked) {
        showError('Please accept the declaration before submitting.');
        return false;
    }

    const submitBtn = document.getElementById('submitBtn');
    const btnText   = document.getElementById('btnText');

    if (submitBtn) {
        submitBtn.disabled = true;
        if (btnText) btnText.innerHTML = '<span class="spinner"></span> Submitting\u2026';
    }

    const formData = {
        school:                          getFieldValue('school'),
        first_name:                      getFieldValue('first_name'),
        second_name:                     getFieldValue('second_name'),
        last_name:                       getFieldValue('last_name'),
        date_of_birth:                   getFieldValue('date_of_birth'),
        gender:                          getFieldValue('gender'),
        student_phone_number:            getFieldValue('student_phone_number'),
        student_class:                   getFieldValue('student_class'),
        student_type:                    getFieldValue('student_type'),
        religion:                        getFieldValue('religion'),
        date_of_admission:               getFieldValue('date_of_admission'),
        national_identification_number:  getFieldValue('national_identification_number'),
        local_identification_number:     getFieldValue('local_identification_number'),
        previous_school_details:         getFieldValue('previous_school_details'),
        medical_history:                 getFieldValue('medical_history'),
        portal_email:                    getFieldValue('portal_email'),
        father_name:                     getFieldValue('father_name'),
        mother_name:                     getFieldValue('mother_name'),
        phone_number:                    getFieldValue('phone_number'),
        mother_phone:                    getFieldValue('mother_phone'),
        father_email:                    getFieldValue('father_email'),
        mother_email:                    getFieldValue('mother_email'),
        father_occupation:               getFieldValue('father_occupation'),
        mother_occupation:               getFieldValue('mother_occupation'),
        if_guardian_is:                  getFieldValue('if_guardian_is'),
        guardian_name:                   getFieldValue('guardian_name'),
        guardian_relation:               getFieldValue('guardian_relation'),
        guardian_phone:                  getFieldValue('guardian_phone'),
        guardian_email:                  getFieldValue('guardian_email'),
        guardian_occupation:             getFieldValue('guardian_occupation'),
        guardian_address:                getFieldValue('guardian_address'),
        current_address:                 getFieldValue('current_address'),
        permanent_address:               getFieldValue('permanent_address'),
        account:                         getFieldValue('account'),
        payment_method:                  getFieldValue('payment_method')
    };

    try {
        // Always get a fresh CSRF token right before submitting
        const csrfToken = await getFrappeCSRF();

        const response = await fetch(
            '/api/method/school.www.student_registration.submit_registration',
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                    'X-Frappe-CSRF-Token': csrfToken
                },
                credentials: 'same-origin',
                body: JSON.stringify({ data: JSON.stringify(formData) })
            }
        );

        // Frappe returns 417 when CSRF fails – give a clear message
        if (response.status === 417) {
            throw new Error('CSRF validation failed (417). Please refresh the page and try again.');
        }

        if (!response.ok) {
            throw new Error('Server error: ' + response.status);
        }

        const result = await response.json();

        if (result.message && result.message.success) {
            document.querySelectorAll('.form-step').forEach(function(step) {
                step.style.display = 'none';
            });
            const successScreen = document.getElementById('successScreen');
            if (successScreen) {
                successScreen.classList.add('show');
                document.getElementById('successMsg').textContent = result.message.message;
                document.getElementById('refBadge').textContent   = 'Reference: ' + result.message.name;
            }
        } else {
            const errorMsg = (result.message && result.message.message) || 'Submission failed. Please try again.';
            showError(errorMsg);
            if (submitBtn) {
                submitBtn.disabled = false;
                if (btnText) btnText.textContent = 'Submit Application';
            }
        }

    } catch (error) {
        console.error('Submission error:', error);
        showError(error.message || 'Network error. Please check your connection and try again.');
        if (submitBtn) {
            submitBtn.disabled = false;
            if (btnText) btnText.textContent = 'Submit Application';
        }
    }

    return false;
}
// ---------------------------------------------------------------------------
// Global exports
// ---------------------------------------------------------------------------

window.loadClassesBySchool    = loadClassesBySchool;
window.toggleGuardianFields   = toggleGuardianFields;
window.submitRegistration     = submitRegistration;
window.goToStep               = goToStep;
window.buildReview            = buildReview;
window.validateForm           = validateForm;

File: school/www/student_registration.py
import frappe
from frappe import _
import json

def get_context(context):
    settings = frappe.get_single("School Settings")

    context.no_cache = 1
    context.show_sidebar = False
    context.no_csrf = 1
    
    # Pass CSRF token to template
    context.csrf_token = frappe.sessions.get_csrf_token()

    context.bill_on_registration = settings.get("bill_on_registration") or 0
    context.require_approval = settings.get("require_approval_before_creating_student") or 0
    
    # Online enrollment setting
    context.allow_online_enrollment = settings.get("allow_online_enrollment") or 0
    context.online_enrollment_enabled = context.allow_online_enrollment

    # Get schools (Cost Centers) with fallback
    schools = frappe.get_all(
        "Cost Center",
        filters={"is_group": 0, "disabled": 0},
        fields=["name", "cost_center_name"]
    )
    if not schools:
        schools = [{"name": "Main Campus", "cost_center_name": "Main Campus"}]
    context.schools = schools

    # Get religions with fallback
    religions = frappe.get_all("Religion", fields=["name"])
    if not religions:
        religions = [{"name": "Christianity"}, {"name": "Islam"}, {"name": "Other"}]
    context.religions = religions

    # Get student categories (Customer Groups)
    context.student_categories = frappe.get_all("Customer Group", fields=["name"])

    # Get houses
    context.houses = frappe.get_all("House", fields=["name"])

    # Get fees structures
    context.fees_structures = frappe.get_all("Fees Structure", fields=["name"])

    # Get categories 1, 2, 3
    context.categories_1 = frappe.get_all("Category 1", fields=["name"])
    context.categories_2 = frappe.get_all("Category 2", fields=["name"])
    context.categories_3 = frappe.get_all("Category 3", fields=["name"])

    # Get area and territory
    context.areas = frappe.get_all("Area", fields=["name"])
    context.territories = frappe.get_all("Territory", fields=["name"])

    # Get fees category
    context.fees_categories = frappe.get_all("Fees Category", fields=["name"])

    # Get payment accounts with fallback
    payment_accounts = frappe.get_all(
        "Account",
        filters={"account_type": ["in", ["Cash", "Bank"]], "is_group": 0, "disabled": 0},
        fields=["name", "account_name"]
    )
    if not payment_accounts:
        payment_accounts = [
            {"name": "Cash - Main", "account_name": "Cash"},
            {"name": "Bank - Main", "account_name": "Bank Account"}
        ]
    context.payment_accounts = payment_accounts

    context.student_types = ["Day", "Boarding"]

    fees_defaults = []
    for row in settings.get("fees_structure_defaults") or []:
        fees_defaults.append({"status": row.status, "fees_structure": row.fees_structure})
    context.fees_defaults = json.dumps(fees_defaults)


@frappe.whitelist(allow_guest=True)
def get_classes_by_school(school):
    """Get classes filtered by school/cost center."""
    if not school:
        return []

    classes = []

    # Approach 1: cost_center field
    if frappe.db.has_column("Student Class", "cost_center"):
        classes = frappe.get_all(
            "Student Class",
            filters={"cost_center": school},
            fields=["name", "class_name"],
            order_by="name"
        )
        if classes:
            return classes

    # Approach 2: school field
    if frappe.db.has_column("Student Class", "school"):
        classes = frappe.get_all(
            "Student Class",
            filters={"school": school},
            fields=["name", "class_name"],
            order_by="name"
        )
        if classes:
            return classes

    # Approach 3: name pattern match
    try:
        classes = frappe.get_all(
            "Student Class",
            filters={"name": ["like", f"%{school}%"]},
            fields=["name", "class_name"],
            order_by="name"
        )
        if classes:
            return classes
    except Exception as e:
        frappe.log_error(f"Error filtering classes by name: {str(e)}", "Student Registration")

    # Approach 4: custom Link field to Cost Center
    try:
        custom_fields = frappe.get_all(
            "Custom Field",
            filters={"dt": "Student Class", "fieldtype": "Link", "options": "Cost Center"},
            fields=["fieldname"]
        )
        for custom_field in custom_fields:
            classes = frappe.get_all(
                "Student Class",
                filters={custom_field.fieldname: school},
                fields=["name", "class_name"],
                order_by="name"
            )
            if classes:
                return classes
    except Exception as e:
        frappe.log_error(f"Error checking custom fields: {str(e)}", "Student Registration")

    # Fallback: return all classes
    return frappe.get_all("Student Class", fields=["name", "class_name"], order_by="name")


@frappe.whitelist(allow_guest=True)
def get_all_student_classes():
    """Get all student classes regardless of school"""
    return frappe.get_all("Student Class", fields=["name", "class_name"], order_by="name")


@frappe.whitelist(allow_guest=True)
def get_sections_by_school(school):
    if not school:
        return []
    if frappe.db.has_column("Section", "cost_center"):
        sections = frappe.get_all("Section", filters={"cost_center": school}, fields=["name"], order_by="name")
        if sections:
            return sections
    if frappe.db.has_column("Section", "school"):
        sections = frappe.get_all("Section", filters={"school": school}, fields=["name"], order_by="name")
        if sections:
            return sections
    return frappe.get_all("Section", fields=["name"], order_by="name")


@frappe.whitelist(allow_guest=True)
def get_sections_by_class_and_school(student_class, school):
    if not student_class:
        return []
    if frappe.db.has_column("Section", "cost_center"):
        sections = frappe.get_all("Section", filters={"name": ["like", f"{student_class}%"], "cost_center": school}, fields=["name"], order_by="name")
        if sections:
            return sections
    if frappe.db.has_column("Section", "school"):
        sections = frappe.get_all("Section", filters={"name": ["like", f"{student_class}%"], "school": school}, fields=["name"], order_by="name")
        if sections:
            return sections
    return frappe.get_all("Section", filters={"name": ["like", f"{student_class}%"]}, fields=["name"], order_by="name")


@frappe.whitelist(allow_guest=True)
def get_all_sections():
    return frappe.get_all("Section", fields=["name"], order_by="name")


@frappe.whitelist(allow_guest=True)
def submit_registration(data):
    """Submit a new student registration"""
    
    # CRITICAL: Disable CSRF for guest users
    frappe.local.no_csrf = True
    
    # Get settings
    settings = frappe.get_single("School Settings")
    
    # Check if online enrollment is enabled
    # Use the exact field name from your School Settings
    if not settings.get("allow_online_enrollment"):
        frappe.throw(_("Online enrollment is currently disabled. Please contact the administrator."))

    if isinstance(data, str):
        data = json.loads(data)

    # Validate required fields
    required = ["first_name", "last_name", "student_class", "student_type", "school"]
    for f in required:
        if not data.get(f):
            frappe.throw(_("Field '{0}' is required.").format(f))

    # Check billing requirements
    if settings.get("bill_on_registration"):
        if not data.get("account"):
            frappe.throw(_("Please select a Payment Account."))
        if not data.get("payment_method"):
            frappe.throw(_("Please select a Payment Method."))

    # Create registration document
    reg = frappe.get_doc({
        "doctype": "Student Online Registration",
        "school": data.get("school"),
        "first_name": data.get("first_name"),
        "second_name": data.get("second_name"),
        "last_name": data.get("last_name"),
        "date_of_birth": data.get("date_of_birth"),
        "gender": data.get("gender"),
        "student_phone_number": data.get("student_phone_number"),
        "student_class": data.get("student_class"),
        "student_type": data.get("student_type"),
        "religion": data.get("religion"),
        "date_of_admission": data.get("date_of_admission"),
        "national_identification_number": data.get("national_identification_number"),
        "local_identification_number": data.get("local_identification_number"),
        "previous_school_details": data.get("previous_school_details"),
        "medical_history": data.get("medical_history"),
        "portal_email": data.get("portal_email"),
        "portal_password": data.get("portal_password"),
        "student_category": data.get("student_category"),
        "house": data.get("house"),
        "has_opening_balance": data.get("has_opening_balance"),
        "opening_balance": data.get("opening_balance"),
        "cost_center": data.get("cost_center"),
        "opening_balance_date": data.get("opening_balance_date"),
        "paying_admin_fee": data.get("paying_admin_fee"),
        "admin_fee_paid": data.get("admin_fee_paid"),
        "admin_fees_structure": data.get("admin_fees_structure"),
        "father_name": data.get("father_name"),
        "mother_name": data.get("mother_name"),
        "phone_number": data.get("phone_number"),
        "mother_phone": data.get("mother_phone"),
        "father_email": data.get("father_email"),
        "mother_email": data.get("mother_email"),
        "father_occupation": data.get("father_occupation"),
        "mother_occupation": data.get("mother_occupation"),
        "if_guardian_is": data.get("if_guardian_is"),
        "guardian_name": data.get("guardian_name"),
        "guardian_relation": data.get("guardian_relation"),
        "guardian_phone": data.get("guardian_phone"),
        "guardian_email": data.get("guardian_email"),
        "guardian_occupation": data.get("guardian_occupation"),
        "guardian_address": data.get("guardian_address"),
        "if_guardian_address_is_current_address": data.get("if_guardian_address_is_current_address"),
        "current_address": data.get("current_address"),
        "if_permanent_address_is_current_address": data.get("if_permanent_address_is_current_address"),
        "permanent_address": data.get("permanent_address"),
        "category_1": data.get("category_1"),
        "category_2": data.get("category_2"),
        "category_3": data.get("category_3"),
        "area": data.get("area"),
        "territory": data.get("territory"),
        "fees_category": data.get("fees_category"),
        "account": data.get("account"),
        "payment_method": data.get("payment_method"),
        "enrollment_status": "Pending"
    })

    reg.insert(ignore_permissions=True)
    frappe.db.commit()

    return {
        "success": True,
        "name": reg.name,
        "status": reg.enrollment_status,
        "message": "Registration submitted successfully! Reference: " + reg.name
    }


@frappe.whitelist(allow_guest=True)
def get_student_classes_with_school_info():
    classes = frappe.get_all("Student Class", fields=["name", "cost_center", "school"])
    custom_fields = frappe.get_all(
        "Custom Field",
        filters={"dt": "Student Class", "fieldtype": "Link", "options": "Cost Center"},
        fields=["fieldname", "label"]
    )
    return {
        "classes": classes,
        "custom_fields": custom_fields,
        "has_cost_center_column": frappe.db.has_column("Student Class", "cost_center"),
        "has_school_column": frappe.db.has_column("Student Class", "school")
    }


@frappe.whitelist(allow_guest=True)
def check_enrollment_status():
    """Helper function to check if online enrollment is enabled"""
    settings = frappe.get_single("School Settings")
    return {
        "enabled": settings.get("allow_online_enrollment") or 0,
        "allow_online_enrollment": settings.get("allow_online_enrollment") or 0
    }
